src/audio/component.c

```diff
@@ -496,6 +496,62 @@ void audio_stream_copy_to_linear(const struct audio_stream *source, int ioffset,
 	}
 }
 
+static inline bool comp_check_eos(struct comp_dev *dev)
+{
+	enum sof_audio_stream_state sink_state = STREAM_STATE_INITIAL;
+	struct comp_buffer *buffer;
+
+	if (!dev->pipeline->expect_eos)
+		return false;
+
+	comp_dev_for_each_producer(dev, buffer) {
+		struct sof_source *source = audio_buffer_get_source(&buffer->audio_buffer);
+		enum sof_audio_stream_state state = source_get_state(source);
+
+		if (source_get_pipeline_id(source) != dev->pipeline->pipeline_id)
+			continue;
```

Comment:
why are producers from the same and from different pipelines treated differently? So this propagation never crosses pipeline borders?